### PR TITLE
Chore tweak operation order to guard against irreversible operations being executed twice

### PR DESCRIPTION
**Pull request by rwrzesien**
Resolves https://github.com/golemfactory/concent/issues/938

### CODE REVIEW

File: concent_api/core/subtask_helpers.py [hunk 1]
```diff
@@ -82,42 +98,41 @@ def finalize_claim_for_acceptance_case() -> None:
             finalize_claim_for_acceptance_case,
             using='control',
         )
-
+    elif subtask.state == Subtask.SubtaskState.VERIFICATION_FILE_TRANSFER.name:  # pylint: disable=no-member
         update_subtask_state(
             subtask=subtask,
-            state=Subtask.SubtaskState.ACCEPTED.name,  # pylint: disable=no-member
```
Comment (pawelkisielewicz) [inline concent_api/core/subtask_helpers.py]: It looks like lot of important things would be changed. It is not, if somebody will review this please use PyCharm to see pull request. Github did it very mistakable, in PyCharm pull request you can see it much better

File: concent_api/core/subtask_helpers.py [hunk 2]
```diff
@@ -70,6 +70,22 @@ def _update_timed_out_subtask(subtask: Subtask) -> None:
...
+            subtask=subtask,
+        )
+        store_pending_message(
+            response_type=PendingResponse.ResponseType.SubtaskResultsSettled,
+            client_public_key=subtask.requestor.public_key_bytes,
+            queue=PendingResponse.Queue.ReceiveOutOfBand,
+            subtask=subtask,
+        )
 
         def finalize_claim_for_acceptance_case() -> None:
```
Comment (pawelkisielewicz) [inline concent_api/core/subtask_helpers.py]: @rwrzesien you moved this section. Why? I don't say it is wrong, but I thought that there is no difference where you put `transaction.on_commit(....`. Does it have to be at the end of `if` or `elif`?

File: concent_api/core/tasks.py [hunk 3]
```diff
@@ -58,6 +58,20 @@ def upload_finished(subtask_id: str) -> None:
         if parse_datetime_to_timestamp(subtask.next_deadline) < get_current_utc_timestamp():
             # Worker makes a payment from requestor's deposit just like in the forced acceptance use case.
 
+            update_subtask_state(
```
Comment (pawelkisielewicz) [inline concent_api/core/tasks.py]: same as above- you moved this block to get `transaction.on_commit(` at the end of function. Is it necessary?

File: concent_api/conductor/service.py [hunk 4]
```diff
@@ -54,3 +59,50 @@ def update_upload_report(file_path: str, result_transfer_request: ResultTransfer
         result_transfer_request.save()
 
         result_upload_finished.delay(result_transfer_request.subtask_id)
+
+
+def store_verification_request_and_blender_subtask_definition(
```
Comment (dybi) [inline concent_api/conductor/service.py]: is there a reason to do both things in one function? Or can it be done sequentially in two separate functions?

Comment (rwrzesien) [reply concent_api/conductor/service.py]: For me there is no reason for this function to exist as it is used only once :)

**Comment by rwrzesien** [comment]
@dybi 

> Side note - I know this code was just moved from one place to another, but I have realized, that some functions don't have unit tests. 

Which ones?

File: concent_api/core/subtask_helpers.py (hunk 2 above)
Comment (rwrzesien) [reply concent_api/core/subtask_helpers.py]: You are right, now it doesn't make a difference, but I have moved it to make it more readable in a way of sentence of operation chain. But if you think it is not necessary I can drop this change.

File: concent_api/core/tasks.py (hunk 3 above)
Comment (rwrzesien) [reply concent_api/core/tasks.py]: Same here, I wanted to make it consistent.

File: concent_api/core/subtask_helpers.py (hunk 1 above)
Comment (rwrzesien) [reply concent_api/core/subtask_helpers.py]: Yes I agree, when I was looking on it in Github I wasn't sure if I am looking at the correct pull request :)

File: concent_api/conductor/views.py [hunk 5]
```diff
@@ -53,13 +53,13 @@ def report_upload(_request: HttpRequest, file_path: str) -> HttpResponse:
...
 
-        # If all expected files have been uploaded, the app sends upload_finished task to the work queue.
-        upload_finished.delay(verification_request.subtask_id)
-
         verification_request.upload_finished = True
         verification_request.full_clean()
         verification_request.save()
 
+        # If all expected files have been uploaded, the app sends upload_finished task to the work queue.
+        upload_finished.delay(verification_request.subtask_id)
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply concent_api/conductor/views.py]: Done, good catch :+1:

**Comment by rwrzesien** [comment]
@pawelkisielewicz 

> Please check line 488 in message_handlers, there is celery task in atomic block. It should be moved away from this block

Done, good catch :+1:

File: concent_api/core/subtask_helpers.py (hunk 2 above)
Comment (pawelkisielewicz) [reply concent_api/core/subtask_helpers.py]: It is OK, thanks for explanation. I was just asking :)

File: concent_api/conductor/views.py (hunk 5 above)
Comment (cameel) [inline concent_api/conductor/views.py]: What about `update_upload_report()` that's called later? It makes database changes too.

BTW, this is not the subject of this pull request but the structure of this function is really confusing and we should fix it, possibly in a separate pull request:
- `update_upload_report()` does not really modify `UploadReport`. It updates `ResultTransferRequest.upload_finished`
- Why is `ResultTransferRequest` modification hidden in a function while for `VerificationRequest` it's all out in the open? These to cases are symmetric: we either deal with an upload request or with a verification request; in both cases we have an upload report, use it to find and update the request, then fire a Celery task. The code should reflect that. I had to spend a few minutes to understand what's actually happening here.

File: concent_api/core/message_handlers.py [hunk 6]
```diff
@@ -989,7 +989,8 @@ def handle_messages_from_database(client_public_key: bytes) -> Union[message.Mes
         elif pending_response.response_type == PendingResponse.ResponseType.ForceReportComputedTaskResponse.name:  # pylint: disable=no-member
             if pending_response.subtask.ack_report_computed_task is not None:
                 ack_report_computed_task = deserialize_message(
-                    pending_response.subtask.ack_report_computed_task.data.tobytes())
+                    pending_response.subtask.ack_report_computed_task.data.tobytes()
+                )
```
Comment (cameel) [inline concent_api/core/message_handlers.py]: All the changes in this file seem unrelated to the commit at all.

File: concent_api/core/subtask_helpers.py (hunk 1 above)
Comment (cameel) [reply concent_api/core/subtask_helpers.py]: That's what happens when you move around a lot of very similar blocks of code. Neither github nor PyCharm is actually right or wrong here. This change can be done both by swapping the whole blocks (as Pycharm probably tells you) or by swapping just the values github shows. They can't know which way you actually did it. That would require understanding the meaning of this code. There are many situations where you'll see the opposite: you changed only a few values but the diff tells you that whole blocks were moved.

File: concent_api/conductor/service.py [hunk 7]
```diff
@@ -54,3 +59,50 @@ def update_upload_report(file_path: str, result_transfer_request: ResultTransfer
         result_transfer_request.save()
 
         result_upload_finished.delay(result_transfer_request.subtask_id)
```
Comment (cameel) [inline concent_api/conductor/service.py]: This task is being posted still inside `@atomic` while the changes have not been committed yet.

File: concent_api/core/message_handlers.py (hunk 6 above)
Comment (rwrzesien) [reply concent_api/core/message_handlers.py]: This is one of the small one-line non-related changes that we discussed on latest retro.

File: concent_api/conductor/views.py (hunk 5 above)
Comment (rwrzesien) [reply concent_api/conductor/views.py]: @cameel 
> update_upload_report() does not really modify UploadReport. It updates ResultTransferRequest.upload_finished

it does with this code:

```
UploadReport.objects.select_for_update().filter(
        path=file_path
    ).update(
        result_transfer_request=result_transfer_request
    )
```

> Why is ResultTransferRequest modification hidden in a function while for VerificationRequest it's all out in the open?

They were done in two different issues with some time interval between them, and `update_upload_report` is reused twice in both view and task.

> What about update_upload_report() that's called later? It makes database changes too.

Yes, but the transaction will still exist when those changes from `update_upload_report` are being done, and after they are all done the task will be scheduled.

File: concent_api/conductor/service.py (hunk 7 above)
Comment (rwrzesien) [reply concent_api/conductor/service.py]: :+1: done
